assets/js/blocks/cart-checkout/cart/edit.js

```diff
@@ -85,11 +86,14 @@ const BlockSettings = ( { attributes, setAttributes } ) => {
 						'woo-gutenberg-products-block'
 					) }
 					checked={ isShippingCalculatorEnabled }
-					onChange={ () =>
+					onChange={ () => {
+						recordEditorEvent( 'cart_shipping_calculator_toggled', {
+							enabled: ! isShippingCalculatorEnabled,
+						} );
```

Comment:
This would have to be implemented inline for everything we want to track right? Right now it's only one attribute, but what if there are many attributes in the same component to track?

One solution would be to use `useEffect` and watch for changes to attributes but that might be tricky to pick out specific changes.

Another idea would be a wrapper for `setAttributes` so that anything using it could fire a tracks event then pass through the changes to `setAttributes`. That way the inline code changes are minimal and all tracking code would be together.

Reply:
setAttribute is essentially a wrapper for setState, if you can possibly hook into the save function of the block or the post content and report from there, this could give more concrete results and could be implemented in the backend (avoid being blocked)

Reply:
I believe that it's best practise to trigger events directly from user actions, as close to the action as possible (I'm looking for guidelines / docs to back up this belief!). 

One risk with watching changes to attributes / state is that you risk firing events that are not tracking user intentions – e.g. default values getting overwritten with real state after a fetch.

Reply:
There is one specific guideline related to this – events should not be fired generically, with the event name dynamically generated (with the exception of prefixing). This is so that you can easily search the code base for an event name, to see how it's fired (for what user action, where in the UI etc).